packages/product/src/services/product-module-service.ts

```diff
@@ -405,6 +398,229 @@ export default class ProductModuleService<
     return products
   }
 
+  @InjectTransactionManager(shouldForceTransaction, "baseRepository_")
+  protected async update_(
+    data: ProductTypes.UpdateProductDTO[],
+    @MedusaContext() sharedContext: Context = {}
+  ): Promise<TProduct[]> {
+    return await this.baseRepository_.transaction(
+      async (manager) => {
+        const productIds = data.map(pd => pd.id)
+        const existingProductVariants = await this.productVariantService_.list(
+          { product_id: productIds },
+          {},
+          sharedContext
+        )
+
+        const existingProductVariantsMap = new Map<
+          string,
+          ProductVariant[]
+        >(
+          data.map((productData) => [
+            productData.id,
+            existingProductVariants.filter((variant) => variant.product_id === productData.id)
+          ])
+        )
+
+        const productVariantsMap = new Map<
+          string,
+          (ProductTypes.CreateProductVariantDTO | ProductTypes.UpdateProductVariantDTO)[]
+        >()
+
+        const productOptionsMap = new Map<
+          string,
+          ProductTypes.CreateProductOptionDTO[]
+        >()
+
+        const productsData = await Promise.all(
+          data.map(async (product) => {
+            const productData = { ...product }
+            const { variants, options } = productData
+            delete productData.options
+            delete productData.variants
+
+            if (!isDefined(productData.id)) {
+              throw new MedusaError(
+                MedusaError.Types.NOT_FOUND,
+                `Cannot update product without id`
+              )
+            }
+
+            productVariantsMap.set(productData.id, variants ?? [])
+            productOptionsMap.set(productData.id, options ?? [])
+
+            if (productData.is_giftcard) {
+              productData.discountable = false
+            }
+
+            await this.setupProductDataForImages(productData, sharedContext)
+            await this.upsertProductTags(productData, sharedContext)
+            await this.upsertProductTypes(productData, sharedContext)
```

Comment:
question: Shouldn't we be doing the same for ProductCollections?

Reply:
as in upsert collections? The core endpoint doesn't seem to be doing that

Reply:
I see. We don't allow you to create collections as part of creating products. Discard this comment :)